File: src/TypeScript/src/numerics/sparse-integer.ts
/** Non-negative arbitrary integer with the SparseInteger arithmetic contract. */
export class SparseInteger {
    readonly #value: bigint;
    public constructor(value: bigint | number | string = 0n) { const parsed = BigInt(value); if (parsed < 0n) throw new RangeError("SparseInteger cannot be negative."); this.#value = parsed; }
    public static parse(text: string): SparseInteger { if (!/^\s*\+?\d+\s*$/.test(text)) throw new SyntaxError("Invalid SparseInteger."); return new SparseInteger(text.trim().replace(/^\+/, "")); }
    public static fromBigInt(value: bigint): SparseInteger { return new SparseInteger(value); }
    public static powerOfTwo(exponent: SparseInteger | bigint | number): SparseInteger { const value = exponent instanceof SparseInteger ? exponent.#value : BigInt(exponent); if (value < 0n || value > BigInt(Number.MAX_SAFE_INTEGER)) throw new RangeError("Exponent cannot be represented by this runtime."); return new SparseInteger(1n << value); }
    public get isZero(): boolean { return this.#value === 0n; }
    public toBigInt(): bigint { return this.#value; }
    public compareTo(other: SparseInteger): number { return this.#value < other.#value ? -1 : this.#value > other.#value ? 1 : 0; }
    public equals(other: SparseInteger): boolean { return this.#value === other.#value; }
    public add(other: SparseInteger): SparseInteger { return new SparseInteger(this.#value + other.#value); }
    public multiply(other: SparseInteger): SparseInteger { return new SparseInteger(this.#value * other.#value); }
    public exactLog2(): SparseInteger { if (this.#value === 0n || (this.#value & (this.#value - 1n)) !== 0n) throw new RangeError("The value is not an exact power of two."); return new SparseInteger(BigInt(this.#value.toString(2).length - 1)); }
    public toString(): string { return this.#value.toString(); }
}

File: src/OCaml/lib/numerics/sparse_integer.ml
type t = Z.t

let of_z value =
  if Z.sign value < 0 then invalid_arg "Sparse_integer cannot be negative";
  value

let of_int value = of_z (Z.of_int value)
let zero = of_int 0
let one = of_int 1
let to_z value = value
let to_string = Z.to_string
let equal = Z.equal
let compare = Z.compare
let is_zero value = Z.equal value Z.zero
let add left right = Z.add left right
let mul left right = Z.mul left right

let parse text =
  let stripped = String.trim text in
  let unsigned =
    if String.length stripped > 0 && stripped.[0] = '+' then
      String.sub stripped 1 (String.length stripped - 1)
    else stripped
  in
  if String.length unsigned = 0 then invalid_arg "invalid Sparse_integer";
  of_z (Z.of_string unsigned)

let power_of_two exponent =
  if not (Z.fits_int exponent) then invalid_arg "exponent is too large";
  Z.shift_left Z.one (Z.to_int exponent)

let power_of_two_int exponent =
  if exponent < 0 then invalid_arg "exponent cannot be negative";
  Z.shift_left Z.one exponent

let exact_log2 value =
  if Z.equal value Z.zero || not (Z.equal (Z.logand value (Z.pred value)) Z.zero) then
    invalid_arg "value is not an exact power of two";
  Z.of_int (Z.numbits value - 1)
